jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,702 @@
+  JLM_ASSERT(is<bitcast_op>(&node));
+
+  const auto & inputRegister = *node.input(0)->origin();
+  const auto & outputRegister = *node.output(0);
+
+  if (!is<PointerType>(inputRegister.type()))
+    return;
+
+  // If the input is a pointer type, the output must also be a pointer type
+  JLM_ASSERT(is<PointerType>(outputRegister.type()));
+
+  const auto inputRegisterPO = Set_->GetRegisterPointerObject(inputRegister);
+  Set_->MapRegisterToExistingPointerObject(outputRegister, inputRegisterPO);
+}
+
+void
+Andersen::AnalyzeBits2ptr(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<bits2ptr_op>(&node));
+  const auto & output = *node.output(0);
+  JLM_ASSERT(is<PointerType>(output.type()));
+
+  // This operation synthesizes a pointer from bytes.
+  // Since no points-to information is tracked through integers, the resulting pointer must
+  // be assumed to possibly point to any external or escaped memory object.
+  const auto outputPO = Set_->CreateRegisterPointerObject(output);
+  Constraints_->AddPointsToExternalConstraint(outputPO);
+}
+
+void
+Andersen::AnalyzeConstantPointerNull(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<ConstantPointerNullOperation>(&node));
+  const auto & output = *node.output(0);
+  JLM_ASSERT(is<PointerType>(output.type()));
+
+  // ConstantPointerNull cannot point to any memory location. We therefore only insert a register
+  // node for it, but let this node not point to anything.
+  Set_->CreateRegisterPointerObject(output);
+}
+
+void
+Andersen::AnalyzeUndef(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<UndefValueOperation>(&node));
+  const auto & output = *node.output(0);
+
+  if (!is<PointerType>(output.type()))
+    return;
+
+  // UndefValue cannot point to any memory location. We therefore only insert a register node for
+  // it, but let this node not point to anything.
+  Set_->CreateRegisterPointerObject(output);
+}
+
+void
+Andersen::AnalyzeMemcpy(const rvsdg::simple_node & node)
+{
+  auto & dstAddressRegister = *node.input(0)->origin();
+  auto & srcAddressRegister = *node.input(1)->origin();
+  JLM_ASSERT(is<PointerType>(dstAddressRegister.type()));
+  JLM_ASSERT(is<PointerType>(srcAddressRegister.type()));
+
+  const auto dstAddressRegisterPO = Set_->GetRegisterPointerObject(dstAddressRegister);
+  const auto srcAddressRegisterPO = Set_->GetRegisterPointerObject(srcAddressRegister);
+
+  // Create an intermediate PointerObject representing the moved values
+  const auto dummyPO = Set_->CreateDummyRegisterPointerObject();
+
+  // Add a "load" constraint from the source into the dummy register
+  Constraints_->AddConstraint(SupersetOfAllPointeesConstraint(dummyPO, srcAddressRegisterPO));
+  // Add a "store" constraint from the dummy register into the destination
+  Constraints_->AddConstraint(AllPointeesPointToSupersetConstraint(dstAddressRegisterPO, dummyPO));
+}
+
+void
+Andersen::AnalyzeConstantArray(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<ConstantStruct>(&node));
```

Comment:
It should be ConstantArray, not ConstantStruct, no? Considering that this not trigger anything, I suspect that there is no unit test that covers this function either?